ls1intum/Artemis
# `Iris`: Integrate with Hestia to generate descriptions for code hints

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


I had the idea to integrate Iris with Hestia a while ago. Iris is now in a state where a first version of this is actually possible.
The goal is to allow editors to automatically generate descriptions for code hints, so that the code hint createion becomes even less work.

### Description

To do this I had to implement a couple things:
- Iris now supports multiple session types
- The old sessions are now chat sessions
- The new sessions for hestia are hestia sessions
- The session service is now supported by two sub services for the respective session types
- The new hestia session service manages the generation of the descriptions

### Steps for Testing

Prerequisites:
- 1 Instructor/Editor
- 1 Java Programming Exercise with Code Coverage enabled
- Only possible on TS9

1. Log in to Artemis
2. Navigate to the programming exercise instructor page
3. Wait for the build plans for solution and template to run through
4. Go to "Hints"
5. Go to "Generate Code Hints"
6. Create the solution code snippets
7. Creat the code hints
8. Edit one of the code hints
9. Press the new button to generate the descriptions
10. Wait a bit
11. The description and content fields should now be filled
12. Save
13. The detail page should show them now
14. If you want view the hint as a student

#### Exam Mode Testing


Hints are not supported for the exam mode

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage




| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| IrisChatSessionService.java | 95% | ✅                           |
| IrisHestiaSessionService.java | 93% | ✅              |
| IrisSessionService.java | 86% | ✅              |
| IrisWebsocketService.java | 88% | ✅              |
| IrisMessageService.java | 100% | ✅              |
| CodeHintService.java | 96% | ✅              |
| exercise-hint-update.component.ts | 84% | ✅              |


### Screenshots


The new button: 
<img width="1252" alt="grafik" src="https://github.com/ls1intum/Artemis/assets/18218285/3d34daee-d5c5-4de4-9983-9d5d8d3ec8f4">

Screencast: 
https://github.com/ls1intum/Artemis/assets/18218285/293f2279-a4a5-4cfb-aede-de70276f5862
